File: js/test/codes_yaml.test.mjs
/**
 * fasten JS codes-yaml tests (P1-11) — node:test, runs in node:20-alpine.
 *
 * Covers the optional yaml catalog: load(), reload() with atomic + fault-
 * tolerant semantics, programmatic-preservation across reload.
 */
import { test, before, beforeEach, describe } from 'node:test';
import assert from 'node:assert/strict';
import fs from 'node:fs/promises';
import path from 'node:path';
import os from 'node:os';

import {
    register, metaOf, registry, codes, _yamlCodes, _resetRegistryForReload,
    _clearRustRegistry,
} from '../src/index.js';
import { _loadedPaths } from '../src/codes_yaml.js';

const FLEET_YAML = `
domain: fleet
emitter: edge-manager
codes:
  FLEET_NODE_REGISTERED:
    category: node.lifecycle
    action: registered
    severity: info
    description: Edge node claimed and registered
  FLEET_DEPLOY_FAILED:
    category: deploy
    action: failed
    severity: error
    retention_class: long
    description: Deployment failed on edge node
`;

let tmpDir;

before(async () => {
    tmpDir = await fs.mkdtemp(path.join(os.tmpdir(), 'fasten-yaml-'));
});

async function writeFile(name, contents) {
    const p = path.join(tmpDir, name);
    await fs.writeFile(p, contents, 'utf8');
    return p;
}

function resetState() {
    _clearRustRegistry();
    registry().clear();
    _yamlCodes.clear();
    _loadedPaths.length = 0;
}

describe('codes_yaml', () => {
    beforeEach(resetState);

    test('load round-trip', async () => {
        const p = await writeFile('round.yaml', FLEET_YAML);
        await codes.load(p);
        const m = metaOf('FLEET_NODE_REGISTERED');
        assert.equal(m.domain, 'fleet');
        assert.equal(m.action, 'registered');
        assert.equal(m.severity, 'info');
        assert.equal(metaOf('FLEET_DEPLOY_FAILED').severity, 'error');
        assert.equal(metaOf('FLEET_DEPLOY_FAILED').retention_class, 'long');
    });

    test('invalid severity raises', async () => {
        const p = await writeFile('badsev.yaml',
            `domain: x\ncodes:\n  BAD_CODE:\n    severity: huge\n`);
        await assert.rejects(() => codes.load(p), /severity='huge'/);
    });

    test('lowercase key raises', async () => {
        const p = await writeFile('badkey.yaml',
            `domain: x\ncodes:\n  lowercase_bad:\n    severity: info\n`);
        await assert.rejects(() => codes.load(p), /UPPER_SNAKE_CASE/);
    });

    test('reload picks up new codes', async () => {
        const p = await writeFile('reload-add.yaml', FLEET_YAML);
        await codes.load(p);
        await fs.writeFile(p, `
domain: fleet
codes:
  FLEET_NODE_REGISTERED:
    severity: info
    description: x
  FLEET_NEW_CODE:
    severity: warn
    description: x
`, 'utf8');
        await codes.reload();
        assert.ok(metaOf('FLEET_NEW_CODE'));
        assert.equal(metaOf('FLEET_NEW_CODE').severity, 'warn');
    });

    test('reload removes codes (full-file replacement)', async () => {
        const p = await writeFile('reload-rm.yaml', FLEET_YAML);
        await codes.load(p);
        assert.ok(metaOf('FLEET_DEPLOY_FAILED'));

        await fs.writeFile(p, `
domain: fleet
codes:
  FLEET_NODE_REGISTERED:
    severity: info
    description: x
`, 'utf8');
        await codes.reload();
        assert.equal(metaOf('FLEET_DEPLOY_FAILED'), undefined);
        assert.ok(metaOf('FLEET_NODE_REGISTERED'));
    });

    test('reload fault-tolerant — bad yaml leaves previous catalog active', async () => {
        const p = await writeFile('reload-bad.yaml', FLEET_YAML);
        await codes.load(p);
        const before = registry().size;

        await fs.writeFile(p, `
domain: fleet
codes:
  FLEET_NODE_REGISTERED:
    severity: completely_bogus
`, 'utf8');
        await assert.rejects(() => codes.reload());

        // Live registry untouched
        assert.equal(registry().size, before);
        assert.ok(metaOf('FLEET_NODE_REGISTERED'));
        assert.ok(metaOf('FLEET_DEPLOY_FAILED'));
    });

    test('reload preserves programmatic codes', async () => {
        const p = await writeFile('reload-prog.yaml', FLEET_YAML);
        await codes.load(p);

        // Programmatic register — different domain, not in yaml
        register('extra', {
            EXTRA_PROG_CODE: { domain: 'extra', category: 'x', action: 'x',
                                severity: 'info', description: 'x', emitter: 't' },
        });
        assert.ok(metaOf('EXTRA_PROG_CODE'));

        await fs.writeFile(p, `
domain: fleet
codes:
  FLEET_NODE_REGISTERED:
    severity: info
    description: x
`, 'utf8');
        await codes.reload();

        assert.ok(metaOf('EXTRA_PROG_CODE'));      // programmatic preserved
        assert.ok(metaOf('FLEET_NODE_REGISTERED')); // yaml reloaded
        assert.equal(metaOf('FLEET_DEPLOY_FAILED'), undefined); // yaml-removed gone
    });
});

File: go/codes_yaml_test.go
package fasten

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

const fleetYAML = `
domain: fleet
emitter: fleet-service
codes:
  FLEET_NODE_REGISTERED:
    category: node.lifecycle
    action: registered
    severity: info
    description: Node claimed and registered
  FLEET_DEPLOY_FAILED:
    category: deploy
    action: failed
    severity: error
    retention_class: long
    description: Deployment failed on node
`

// resetYAMLState clears yaml-loader state between tests.
func resetYAMLState(t *testing.T) {
	t.Helper()
	yamlMu.Lock()
	yamlPaths = nil
	yamlCodes = map[Code]bool{}
	yamlMu.Unlock()
	clearBothRegistries()
	t.Cleanup(func() {
		yamlMu.Lock()
		yamlPaths = nil
		yamlCodes = map[Code]bool{}
		yamlMu.Unlock()
		clearBothRegistries()
	})
}

func writeYAML(t *testing.T, contents string) string {
	t.Helper()
	dir := t.TempDir()
	path := filepath.Join(dir, "fleet.codes.yaml")
	if err := os.WriteFile(path, []byte(contents), 0o644); err != nil {
		t.Fatalf("writeYAML: %v", err)
	}
	return path
}

func TestLoad_RoundTrip(t *testing.T) {
	resetYAMLState(t)
	path := writeYAML(t, fleetYAML)
	if err := Load(path); err != nil {
		t.Fatalf("Load: %v", err)
	}
	m, ok := metaOf("FLEET_NODE_REGISTERED")
	if !ok {
		t.Fatal("FLEET_NODE_REGISTERED not in registry")
	}
	if m.Domain != "fleet" || m.Action != "registered" || m.Severity != SevInfo {
		t.Fatalf("unexpected meta: %+v", m)
	}
	m2, _ := metaOf("FLEET_DEPLOY_FAILED")
	if m2.Severity != SevError || m2.RetentionClass != RetLong {
		t.Fatalf("unexpected meta: %+v", m2)
	}
}

func TestLoad_InvalidSeverity(t *testing.T) {
	resetYAMLState(t)
	path := writeYAML(t, `
domain: x
codes:
  BAD_CODE:
    severity: huge
`)
	err := Load(path)
	if err == nil || !strings.Contains(err.Error(), `severity="huge"`) {
		t.Fatalf("expected severity error, got %v", err)
	}
}

func TestLoad_LowercaseKey(t *testing.T) {
	resetYAMLState(t)
	path := writeYAML(t, `
domain: x
codes:
  lowercase_bad:
    severity: info
`)
	err := Load(path)
	if err == nil || !strings.Contains(err.Error(), "UPPER_SNAKE_CASE") {
		t.Fatalf("expected key-shape error, got %v", err)
	}
}

func TestReload_RemovesCodes(t *testing.T) {
	resetYAMLState(t)
	path := writeYAML(t, fleetYAML)
	if err := Load(path); err != nil {
		t.Fatalf("Load: %v", err)
	}
	if _, ok := metaOf("FLEET_DEPLOY_FAILED"); !ok {
		t.Fatal("FLEET_DEPLOY_FAILED missing pre-reload")
	}

	// Rewrite the file: drop FLEET_DEPLOY_FAILED.
	trimmed := `
domain: fleet
codes:
  FLEET_NODE_REGISTERED:
    category: node.lifecycle
    action: registered
    severity: info
    description: x
`
	if err := os.WriteFile(path, []byte(trimmed), 0o644); err != nil {
		t.Fatalf("rewrite: %v", err)
	}
	if err := Reload(); err != nil {
		t.Fatalf("Reload: %v", err)
	}
	if _, ok := metaOf("FLEET_DEPLOY_FAILED"); ok {
		t.Fatal("FLEET_DEPLOY_FAILED should be gone after reload")
	}
	if _, ok := metaOf("FLEET_NODE_REGISTERED"); !ok {
		t.Fatal("FLEET_NODE_REGISTERED missing after reload")
	}
}

func TestReload_FaultTolerant(t *testing.T) {
	resetYAMLState(t)
	path := writeYAML(t, fleetYAML)
	if err := Load(path); err != nil {
		t.Fatalf("Load: %v", err)
	}
	// Snapshot known-good registry contents
	regMu.RLock()
	snapshot := make(map[Code]Meta, len(_registry))
	for k, v := range _registry {
		snapshot[k] = v
	}
	regMu.RUnlock()

	// Corrupt the yaml — invalid severity
	if err := os.WriteFile(path, []byte(`
domain: fleet
codes:
  FLEET_NODE_REGISTERED:
    severity: completely_bogus
`), 0o644); err != nil {
		t.Fatalf("corrupt write: %v", err)
	}
	if err := Reload(); err == nil {
		t.Fatal("expected reload to error on bad severity")
	}

	// Live registry must be unchanged.
	regMu.RLock()
	defer regMu.RUnlock()
	if len(_registry) != len(snapshot) {
		t.Fatalf("registry mutated by failed reload: before=%d after=%d", len(snapshot), len(_registry))
	}
	for k, v := range snapshot {
		if got, ok := _registry[k]; !ok || got.Severity != v.Severity {
			t.Fatalf("registry mutated by failed reload: key=%s", k)
		}
	}
}

func TestReload_PreservesProgrammatic(t *testing.T) {
	resetYAMLState(t)
	path := writeYAML(t, fleetYAML)
	if err := Load(path); err != nil {
		t.Fatalf("Load: %v", err)
	}

	// Add a programmatic code (different domain, not in yaml)
	if err := Register("extra", map[Code]Meta{
		"EXTRA_PROG_CODE": {Domain: "extra", Category: "x", Action: "x",
			Severity: SevInfo, Description: "x", Emitter: "t"},
	}); err != nil {
		t.Fatalf("Register: %v", err)
	}

	// Modify yaml + reload
	if err := os.WriteFile(path, []byte(`
domain: fleet
codes:
  FLEET_NODE_REGISTERED:
    severity: info
    description: x
`), 0o644); err != nil {
		t.Fatalf("rewrite: %v", err)
	}
	if err := Reload(); err != nil {
		t.Fatalf("Reload: %v", err)
	}

	if _, ok := metaOf("EXTRA_PROG_CODE"); !ok {
		t.Fatal("programmatic EXTRA_PROG_CODE lost across reload")
	}
	if _, ok := metaOf("FLEET_NODE_REGISTERED"); !ok {
		t.Fatal("FLEET_NODE_REGISTERED lost across reload")
	}
	if _, ok := metaOf("FLEET_DEPLOY_FAILED"); ok {
		t.Fatal("FLEET_DEPLOY_FAILED should be removed by reload")
	}
}

func TestReload_NoOpWithoutLoad(t *testing.T) {
	resetYAMLState(t)
	if err := Reload(); err != nil {
		t.Fatalf("Reload with no prior Load should be no-op, got %v", err)
	}
}
